core/src/serve_repair.rs

```diff
@@ -592,31 +940,134 @@ impl ServeRepair {
+                sender: self.my_id(),
+                recipient: *repair_peer_id,
+                timestamp: timestamp(),
+                nonce,
+            })
+        } else {
+            None
+        };
+        let request_proto = match repair_request {
             ShredRepairType::Shred(slot, shred_index) => {
                 repair_stats
                     .shred
                     .update(repair_peer_id, *slot, *shred_index);
-                Ok(self.window_index_request_bytes(*slot, *shred_index, nonce)?)
+                if let Some(header) = header {
+                    RepairProtocol::WindowIndex {
+                        header,
+                        slot: *slot,
+                        shred_index: *shred_index,
+                    }
+                } else {
+                    RepairProtocol::LegacyWindowIndexWithNonce(
+                        self.my_info(),
+                        *slot,
+                        *shred_index,
+                        nonce,
+                    )
+                }
             }
             ShredRepairType::HighestShred(slot, shred_index) => {
                 repair_stats
                     .highest_shred
                     .update(repair_peer_id, *slot, *shred_index);
-                Ok(self.window_highest_index_request_bytes(*slot, *shred_index, nonce)?)
+                if let Some(header) = header {
+                    RepairProtocol::HighestWindowIndex {
+                        header,
+                        slot: *slot,
+                        shred_index: *shred_index,
+                    }
+                } else {
+                    RepairProtocol::LegacyHighestWindowIndexWithNonce(
+                        self.my_info(),
+                        *slot,
+                        *shred_index,
+                        nonce,
+                    )
+                }
             }
             ShredRepairType::Orphan(slot) => {
                 repair_stats.orphan.update(repair_peer_id, *slot, 0);
-                Ok(self.orphan_bytes(*slot, nonce)?)
+                if let Some(header) = header {
+                    RepairProtocol::Orphan {
+                        header,
+                        slot: *slot,
+                    }
+                } else {
+                    RepairProtocol::LegacyOrphanWithNonce(self.my_info(), *slot, nonce)
+                }
+            }
+        };
+        Self::repair_proto_to_bytes(&request_proto, identity_keypair)
+    }
+
+    /// Distinguish and process `RepairResponse` ping packets ignoring other
+    /// packets in the batch.
+    pub(crate) fn handle_repair_response_pings(
+        repair_socket: &UdpSocket,
+        keypair: &Keypair,
+        packet_batch: &mut PacketBatch,
+        stats: &mut ShredFetchStats,
+    ) {
+        let mut pending_pongs = Vec::default();
+        for packet in packet_batch.iter_mut() {
+            if packet.meta.size != REPAIR_RESPONSE_SERIALIZED_PING_BYTES {
```

Comment:
@jbiseda Looking at this again I am not sure this will work.
Seems like
```rust
REPAIR_RESPONSE_SERIALIZED_PING_BYTES = 132
```
But data shred headers are 88 bytes so we can have a data shred which is the same size of as the ping packet.

Maybe we should `packet.meta.set_discard(true)` only if `ping.verfiy()` is true;
also confirm that no valid shred can be of the same size and deserialize as a `Ping` and verify.

Reply:
See #26878